experiments/exp163_bert.py

```diff
@@ -0,0 +1,316 @@
+)
+from marin.processing.classification.inference import InferenceConfig, run_inference
+from marin.processing.tokenize.data_configs import lm_mixture_data_config
+
+dolmino_dclm = get_dolmino_step_llama3("dclm")
+
+
+@dataclass
+class ExperimentConfig:
+    """Configuration for comparing BERT vs Fasttext classifiers.
+
+    This config defines parameters for an experiment that:
+    1. Takes input documents from specified data sources
+    2. Takes positive and negative documents from specified data sources.
+    3. Trains both a BERT and a Fasttext classifier on the positive/negative documents.
+    4. Filters and tokenizes the resulting data.
+
+    Args:
+        experiment_name: Identifier for this experiment
+        input_data_source_to_path: Mapping of data source names to their GCS paths
+        keep_fraction: Fraction of highest-quality documents to keep after filtering
+    """
+
+    experiment_name: str
+    classifier_training_datasets: list[DatasetConfig]
+    input_data_source_to_path: dict[str, str] = field(
+        default_factory=lambda: {
+            "fineweb_2024_18": (
+                "gs://marin-us-central2/documents/fineweb-small-resiliparse-preserve-formatting-v2-e72837/md/CC-MAIN-2024-18/"
+            ),
+        }
+    )
+    keep_fractions: list[float] = field(default_factory=lambda: [0.01, 0.05, 0.1, 0.2])
+
+
+def create_steps(config: ExperimentConfig) -> list[ExecutorStep]:
+    """Create the steps for a single experiment.
+
+    Variation of exp614_quality_filtering.py, but trains both BERT and Fasttext quality classifiers.
+    """
+
+    steps = []
+
+    fasttext_classifier_train = ExecutorStep(
+        name=f"classifiers/{config.experiment_name}/fasttext",
+        fn=train_fasttext,
+        config=TrainFasttextClassifierConfig(
+            datasets=config.classifier_training_datasets,
+            output_path=this_output_path(),
+            fasttext_args={"lr": versioned(0.1), "thread": 4, "wordNgrams": 2},
+            val_frac=versioned(0.1),
+            seed=versioned(0),
+        ),
+        pip_dependency_groups=["fasttext"],
+    )
+
+    bert_classifier_train = ExecutorStep(
+        name=f"classifiers/{config.experiment_name}/bert-debug",  # TODO: remove debug
+        fn=train_bert,
+        config=TrainBertClassifierConfig(
+            datasets=config.classifier_training_datasets,
+            output_path=this_output_path(),
+            val_frac=versioned(0.1),
+            seed=versioned(0),
+        ),
+        pip_dependency_groups=[
+            "--find-links https://storage.googleapis.com/libtpu-releases/index.html",
+            "--find-links https://storage.googleapis.com/libtpu-wheels/index.html",
+            "fasttext",
+            "datasets",
+            "filelock",
+            "torch",
+            "torch_xla[tpu]",
+            "accelerate",
+        ],
+    )
+
+    for input_data_source, input_data_path in config.input_data_source_to_path.items():
```

Comment:
I'd probably break this function up...you can have one function that trains models given an input source, and then another function that takes another an input source (and the corresponding classifiers, keep fraction), to avoid large blocks of code with deep nesting.  Then this function that has the nesting can be quite small.